File: src/script_generator.py
"""
Generador automático de guiones para videos de Cine Norte
"""

import openai
import json
import re
from typing import Dict, List, Optional
from dataclasses import dataclass
import logging
from datetime import datetime

from config import API_KEYS, CONTENT_CONFIG, BRANDING
from src.content_analyzer import ContentItem

logger = logging.getLogger(__name__)

@dataclass
class ScriptSegment:
    """Segmento del guion con timing y elementos visuales"""
    text: str
    start_time: float
    end_time: float
    visual_cues: List[str]
    emphasis_words: List[str]
    background_music: str

@dataclass
class GeneratedScript:
    """Guion completo generado"""
    title: str
    content: ContentItem
    segments: List[ScriptSegment]
    total_duration: float
    hashtags: List[str]
    description: str
    thumbnail_prompts: List[str]
    raw_text: str

class ScriptGenerator:
    """Generador de guiones automático con IA"""
    
    def __init__(self):
        self.openai_api_key = API_KEYS.get("openai")
        if self.openai_api_key:
            openai.api_key = self.openai_api_key
        
        # Plantillas de guion por tipo de contenido
        self.script_templates = {
            "movie": {
                "intro": "¡Hola cinéfilos! Soy Cine Norte y hoy les traigo un análisis de {title}",
                "structure": [
                    "intro_hook",
                    "plot_summary", 
                    "key_scenes",
                    "technical_aspects",
                    "verdict",
                    "outro"
                ]
            },
            "tv": {
                "intro": "¡Bienvenidos a Cine Norte! Hoy analizamos {title}",
                "structure": [
                    "intro_hook",
                    "series_overview",
                    "season_highlights",
                    "character_analysis", 
                    "verdict",
                    "outro"
                ]
            }
        }
    
    def generate_script(self, content: ContentItem, style: str = "engaging") -> GeneratedScript:
        """
        Genera un guion completo para el contenido
        
        Args:
            content: Información del contenido
            style: Estilo del guion ('engaging', 'dramatic', 'informative')
        """
        try:
            # Generar texto base del guion
            script_text = self._generate_script_text(content, style)
            
            # Dividir en segmentos con timing
            segments = self._create_script_segments(script_text, content)
            
            # Generar metadatos
            hashtags = self._generate_hashtags(content)
            description = self._generate_description(content, script_text)
            thumbnail_prompts = self._generate_thumbnail_prompts(content)
            
            # Calcular duración total
            total_duration = sum(segment.end_time - segment.start_time for segment in segments)
            
            return GeneratedScript(
                title=f"Análisis de {content.title} - Cine Norte",
                content=content,
                segments=segments,
                total_duration=total_duration,
                hashtags=hashtags,
                description=description,
                thumbnail_prompts=thumbnail_prompts,
                raw_text=script_text
            )
            
        except Exception as e:
            logger.error(f"Error generando guion: {e}")
            return self._create_fallback_script(content)
    
    def _generate_script_text(self, content: ContentItem, style: str) -> str:
        """Genera el texto base del guion usando OpenAI"""
        try:
            if not self.openai_api_key:
                return self._generate_fallback_script(content)
            
            # Construir prompt contextual
            prompt = self._build_script_prompt(content, style)
            
            response = openai.ChatCompletion.create(
                model="gpt-4",
                messages=[
                    {
                        "role": "system",
                        "content": "Eres un experto en análisis cinematográfico y creador de contenido para YouTube. Tu estilo es dinámico, entretenido y sin spoilers importantes. Eres parte del canal Cine Norte."
                    },
                    {
                        "role": "user", 
                        "content": prompt
                    }
                ],
                max_tokens=800,
                temperature=0.7
            )
            
            return response.choices[0].message.content.strip()
            
        except Exception as e:
            logger.error(f"Error con OpenAI: {e}")
            return self._generate_fallback_script(content)
    
    def _build_script_prompt(self, content: ContentItem, style: str) -> str:
        """Construye el prompt para la generación del guion"""
        
        # Determinar duración objetivo
        target_duration = min(CONTENT_CONFIG["max_script_length"], 400)
        
        # Estructura base
        structure = self.script_templates.get(content.content_type, self.script_templates["movie"])
        
        prompt = f"""
Genera un guion de video para YouTube de máximo {target_duration} palabras sobre:

TÍTULO: {content.title}
TIPO: {'Película' if content.content_type == 'movie' else 'Serie'}
GÉNEROS: {', '.join(content.genres)}
SINOPSIS: {content.overview}
RATING: {content.rating}/10
DURACIÓN OBJETIVO: 2-3 minutos de video

ESTILO: {style}
CANAL: Cine Norte (análisis cinematográfico en español)

ESTRUCTURA REQUERIDA:
1. Hook inicial (10-15 segundos) - Captar atención inmediatamente
2. Resumen de la trama (30-45 segundos) - Sin spoilers importantes
3. Análisis de aspectos destacados (45-60 segundos) - Actuaciones, dirección, efectos
4. Veredicto personal (20-30 segundos) - Recomendación honesta
5. Cierre con call-to-action (10-15 segundos)

REGLAS IMPORTANTES:
- Usar un tono dinámico y entretenido
- Incluir frases impactantes para el texto en pantalla
- Evitar spoilers importantes
- Mencionar elementos técnicos relevantes
- Incluir referencias a otras películas/series cuando sea apropiado
- Usar lenguaje coloquial pero profesional
- Incluir transiciones fluidas entre secciones

FORMATO: Escribe el guion completo como un solo texto fluido, sin marcas de tiempo ni indicaciones técnicas.
"""
        
        return prompt
    
    def _create_script_segments(self, script_text: str, content: ContentItem) -> List[ScriptSegment]:
        """Divide el guion en segmentos con timing y elementos visuales"""
        segments = []
        
        # Dividir el texto en párrafos
        paragraphs = [p.strip() for p in script_text.split('\n\n') if p.strip()]
        
        current_time = 0.0
        words_per_minute = 150  # Velocidad de narración
        
        for i, paragraph in enumerate(paragraphs):
            # Calcular duración basada en palabras
            word_count = len(paragraph.split())
            duration = (word_count / words_per_minute) * 60
            
            # Generar elementos visuales
            visual_cues = self._generate_visual_cues(paragraph, content, i)
            emphasis_words = self._extract_emphasis_words(paragraph)
            background_music = self._select_background_music(content, i, len(paragraphs))
            
            segment = ScriptSegment(
                text=paragraph,
                start_time=current_time,
                end_time=current_time + duration,
                visual_cues=visual_cues,
                emphasis_words=emphasis_words,
                background_music=background_music
            )
            
            segments.append(segment)
            current_time += duration
        
        return segments
    
    def _generate_visual_cues(self, text: str, content: ContentItem, segment_index: int) -> List[str]:
        """Genera indicaciones visuales para el segmento"""
        cues = []
        
        # Cues basadas en contenido del texto
        if "acción" in text.lower() or "explosión" in text.lower():
            cues.append("transición_energética")
            cues.append("efectos_visuales_dinámicos")
        
        if "suspense" in text.lower() or "misterio" in text.lower():
            cues.append("iluminación_tenue")
            cues.append("movimiento_cámara_suave")
        
        if "drama" in text.lower() or "emotivo" in text.lower():
            cues.append("primer_plano")
            cues.append("colores_saturados")
        
        # Cues basadas en el segmento
        if segment_index == 0:  # Intro
            cues.extend(["logo_cine_norte", "efecto_revelador"])
        elif segment_index == len(text.split('\n\n')) - 1:  # Outro
            cues.extend(["logo_cine_norte", "call_to_action"])
        
        # Cues específicas del contenido
        if content.content_type == "movie":
            cues.append("poster_película")
        else:
            cues.append("poster_serie")
        
        return cues
    
    def _extract_emphasis_words(self, text: str) -> List[str]:
        """Extrae palabras clave para enfatizar visualmente"""
        # Palabras de impacto común en análisis cinematográfico
        impact_words = [
            "increíble", "espectacular", "sorprendente", "impactante",
            "brillante", "genial", "perfecto", "excelente", "magnífico",
            "terrible", "decepcionante", "aburrido", "confuso"
        ]
        
        words = text.lower().split()
        emphasis = [word for word in words if any(impact in word for impact in impact_words)]
        
        return emphasis[:5]  # Máximo 5 palabras de énfasis
    
    def _select_background_music(self, content: ContentItem, segment_index: int, total_segments: int) -> str:
        """Selecciona música de fondo apropiada"""
        # Música basada en género
        genre_music = {
            "Acción": "epic_action",
            "Terror": "dark_suspense", 
            "Drama": "emotional_drama",
            "Comedia": "light_upbeat",
            "Ciencia ficción": "futuristic_synth"
        }
        
        # Música basada en el segmento
        if segment_index == 0:
            return "intro_energetic"
        elif segment_index == total_segments - 1:
            return "outro_motivational"
        else:
            # Seleccionar basado en el primer género
            if content.genres:
                first_genre = content.genres[0]
                return genre_music.get(first_genre, "neutral_cinematic")
            return "neutral_cinematic"
    
    def _generate_hashtags(self, content: ContentItem) -> List[str]:
        """Genera hashtags optimizados para SEO"""
        hashtags = []
        
        # Hashtags base
        base_tags = ["#CineNorte", "#AnálisisCinematográfico", "#Reseña"]
        hashtags.extend(base_tags)
        
        # Hashtags por tipo
        if content.content_type == "movie":
            hashtags.extend(["#Película", "#Cine", "#ReseñaPelicula"])
        else:
            hashtags.extend(["#Serie", "#TV", "#ReseñaSerie"])
        
        # Hashtags por género
        for genre in content.genres[:3]:  # Máximo 3 géneros
            genre_tag = f"#{genre.replace(' ', '')}"
            hashtags.append(genre_tag)
        
        # Hashtags por plataforma (si está disponible)
        for platform in content.platforms[:2]:  # Máximo 2 plataformas
            platform_tag = f"#{platform.replace(' ', '').replace('+', 'Plus')}"
            hashtags.append(platform_tag)
        
        # Hashtags de tendencia
        trending_tags = ["#Streaming", "#Entretenimiento", "#CulturaPop"]
        hashtags.extend(trending_tags)
        
        return hashtags[:CONTENT_CONFIG.get("max_hashtags", 10)]
    
    def _generate_description(self, content: ContentItem, script_text: str) -> str:
        """Genera descripción para YouTube"""
        # Resumen corto del contenido
        summary = content.overview[:100] + "..." if len(content.overview) > 100 else content.overview
        
        description = f"""
🎬 ANÁLISIS COMPLETO: {content.title}

{summary}

⭐ RATING: {content.rating}/10
🎭 GÉNEROS: {', '.join(content.genres)}
📅 ESTRENO: {content.release_date}

{script_text[:200]}...

🔔 ¡SUSCRÍBETE para más análisis cinematográficos!
👍 ¡DALE LIKE si te gustó el video!
💬 ¡COMENTA tu opinión sobre {content.title}!

{', '.join(self._generate_hashtags(content))}

---
Cine Norte - Tu canal de análisis cinematográfico
"""
        
        return description
    
    def _generate_thumbnail_prompts(self, content: ContentItem) -> List[str]:
        """Genera prompts para crear miniaturas atractivas"""
        prompts = []
        
        # Prompt principal
        main_prompt = f"""
Cinematic thumbnail for {content.title} movie review:
- {content.title} text in bold red letters
- Dark cinematic background with {content.genres[0] if content.genres else 'dramatic'} atmosphere
- Professional movie poster style
- Cine Norte logo visible
- High contrast, eye-catching design
- {content.rating}/10 rating displayed
"""
        prompts.append(main_prompt)
        
        # Prompts alternativos
        if content.content_type == "movie":
            prompts.append(f"Movie poster style thumbnail: {content.title} with explosive action elements")
            prompts.append(f"Dramatic close-up style: {content.title} with intense lighting")
        else:
            prompts.append(f"TV series style thumbnail: {content.title} with character portraits")
            prompts.append(f"Season poster style: {content.title} with ensemble cast")
        
        return prompts
    
    def _generate_fallback_script(self, content: ContentItem) -> str:
        """Genera un guion básico sin IA como respaldo"""
        script = f"""
¡Hola cinéfilos! Soy Cine Norte y hoy les traigo un análisis de {content.title}.

{content.overview}

Esta {'película' if content.content_type == 'movie' else 'serie'} de {', '.join(content.genres[:2])} 
nos presenta una historia que mantiene al espectador en vilo desde el primer momento.

Los aspectos técnicos están muy bien logrados, con una dirección sólida y actuaciones convincentes. 
La cinematografía captura perfectamente la atmósfera que la historia requiere.

En general, {content.title} es una {'película' if content.content_type == 'movie' else 'serie'} 
que vale la pena ver, especialmente si eres fan del género {content.genres[0] if content.genres else 'drama'}.

¿Qué opinas de {content.title}? Déjamelo en los comentarios y no olvides suscribirte para más análisis cinematográficos.
"""
        return script
    
    def _create_fallback_script(self, content: ContentItem) -> GeneratedScript:
        """Crea un guion de respaldo en caso de error"""
        script_text = self._generate_fallback_script(content)
        segments = self._create_script_segments(script_text, content)
        
        return GeneratedScript(
            title=f"Análisis de {content.title} - Cine Norte",
            content=content,
            segments=segments,
            total_duration=sum(segment.end_time - segment.start_time for segment in segments),
            hashtags=self._generate_hashtags(content),
            description=self._generate_description(content, script_text),
            thumbnail_prompts=self._generate_thumbnail_prompts(content),
            raw_text=script_text
        )
    
    def save_script_to_file(self, script: GeneratedScript, filename: str = None) -> str:
        """Guarda el guion en un archivo de texto"""
        if not filename:
            timestamp = datetime.now().strftime("%Y%m%d_%H%M%S")
            filename = f"guion_{script.content.title.replace(' ', '_')}_{timestamp}.txt"
        
        filepath = f"output/{filename}"
        
        with open(filepath, 'w', encoding='utf-8') as f:
            f.write(f"GUION: {script.title}\n")
            f.write("=" * 50 + "\n\n")
            f.write(script.raw_text)
            f.write("\n\n" + "=" * 50 + "\n")
            f.write("HASHTAGS:\n")
            f.write(", ".join(script.hashtags))
            f.write("\n\nDESCRIPCIÓN:\n")
            f.write(script.description)
        
        return filepath

File: js/script-generator.js
/**
 * Generador de Guiones con IA para Cine Norte
 * Crea guiones optimizados para videos de redes sociales
 */

class ScriptGenerator {
    constructor() {
        this.openaiBaseUrl = 'https://api.openai.com/v1';
        this.cache = new Map();
    }
    
    /**
     * Genera un guion completo para el contenido
     * @param {Object} content - Información del contenido
     * @param {string} targetPlatform - Plataforma objetivo
     * @param {number} durationTarget - Duración objetivo en segundos
     * @param {string} style - Estilo del guion
     * @returns {Promise<Object>} Guion generado
     */
    async generateScript(content, targetPlatform = 'youtube', durationTarget = 120, style = 'dynamic') {
        try {
            const apiKey = config.get('apis.openai.key');
            if (!apiKey) {
                throw new Error('OpenAI API key no configurada');
            }
            
            // Crear prompt base
            const prompt = this._createBasePrompt(content, targetPlatform, durationTarget, style);
            
            // Generar guion con OpenAI
            const response = await fetch(`${this.openaiBaseUrl}/chat/completions`, {
                method: 'POST',
                headers: {
                    'Authorization': `Bearer ${apiKey}`,
                    'Content-Type': 'application/json'
                },
                body: JSON.stringify({
                    model: 'gpt-4',
                    messages: [
                        {
                            role: 'system',
                            content: this._getSystemPrompt()
                        },
                        {
                            role: 'user',
                            content: prompt
                        }
                    ],
                    max_tokens: 2000,
                    temperature: 0.7
                })
            });
            
            if (!response.ok) {
                const error = await response.json();
                throw new Error(`Error OpenAI: ${error.error?.message || 'Error desconocido'}`);
            }
            
            const data = await response.json();
            const scriptText = data.choices[0].message.content;
            
            // Procesar y estructurar el guion
            const sections = this._parseScriptSections(scriptText, content);
            
            // Generar metadatos adicionales
            const hashtags = config.generateHashtags(content, targetPlatform);
            const titleSuggestions = this._generateTitleSuggestions(content, targetPlatform);
            const visualStyle = this._determineVisualStyle(content, style);
            const musicSuggestion = this._suggestMusic(content, style);
            
            // Calcular duración total
            const totalDuration = sections.reduce((sum, section) => sum + section.duration_seconds, 0);
            
            return {
                title: content.title,
                content: content,
                sections: sections,
                total_duration: totalDuration,
                word_count: scriptText.split(' ').length,
                target_platform: targetPlatform,
                hashtags: hashtags,
                title_suggestions: titleSuggestions,
                visual_style: visualStyle,
                music_suggestion: musicSuggestion,
                raw_text: scriptText,
                generated_at: new Date().toISOString()
            };
            
        } catch (error) {
            console.error('Error generando guion:', error);
            return this._createFallbackScript(content, targetPlatform);
        }
    }
    
    /**
     * Crea el prompt base para la generación
     * @param {Object} content - Información del contenido
     * @param {string} targetPlatform - Plataforma objetivo
     * @param {number} durationTarget - Duración objetivo
     * @param {string} style - Estilo del guion
     * @returns {string} Prompt completo
     */
    _createBasePrompt(content, targetPlatform, durationTarget, style) {
        const platformInstructions = {
            youtube: 'Optimizado para YouTube: guion más largo, análisis detallado, intros/outros elaborados',
            tiktok: 'Optimizado para TikTok: guion corto y directo, hooks inmediatos, ritmo acelerado',
            instagram: 'Optimizado para Instagram: visual, estético, historias en highlights',
            facebook: 'Optimizado para Facebook: contenido compartible, engagement alto',
            twitter: 'Optimizado para Twitter: conciso, impactante, fácil de compartir'
        };
        
        const styleInstructions = {
            dynamic: 'Estilo dinámico: ritmo rápido, cortes frecuentes, mucha energía',
            dramatic: 'Estilo dramático: pausas estratégicas, énfasis emocional, tensión',
            comedic: 'Estilo cómico: humor sutil, comentarios ingeniosos, tono ligero',
            analytical: 'Estilo analítico: enfoque en técnica, dirección, actuaciones'
        };
        
        return `
Crea un guion para un video de Cine Norte sobre:

TÍTULO: ${content.title}
TIPO: ${content.content_type.toUpperCase()}
GÉNERO: ${content.genre.join(', ')}
PLATAFORMA: ${content.platform}
RATING: ${content.rating}/10
SINOPSIS: ${content.overview}
FECHA DE ESTRENO: ${content.release_date}
DURACIÓN OBJETIVO: ${durationTarget} segundos

INSTRUCCIONES ESPECÍFICAS:
- Plataforma: ${platformInstructions[targetPlatform] || 'General'}
- Estilo: ${styleInstructions[style] || 'Dinámico'}
- Duración máxima: 3 minutos
- Idioma: Español (México)
- Sin spoilers importantes

ESTRUCTURA REQUERIDA:
1. INTRO (5-10 seg): Hook inmediato + presentación Cine Norte
2. HOOK (10-15 seg): Elemento más atractivo sin spoilers
3. PLOT (60-90 seg): Resumen de trama con análisis
4. ANÁLISIS (30-45 seg): Aspectos técnicos o temáticos
5. OUTRO (10-15 seg): Call-to-action + despedida

Para cada sección incluye:
- Texto del guion
- Duración estimada en segundos
- Indicaciones visuales específicas
- Palabras clave para énfasis
- Emoción objetivo

Formato de respuesta en JSON estructurado.
        `.trim();
    }
    
    /**
     * Obtiene el prompt del sistema para la IA
     * @returns {string} Prompt del sistema
     */
    _getSystemPrompt() {
        return `
Eres un experto editor audiovisual y creador de contenido para Cine Norte, 
una marca especializada en análisis cinematográfico para redes sociales.

Tu misión es crear guiones dinámicos, atractivos y optimizados para engagement
que mantengan la identidad de marca de Cine Norte.

Características de Cine Norte:
- Paleta de colores: Rojo #E50914, Negro #0A0A0A, Plateado #C0C0C0
- Estilo: Cinematográfico, profesional, con toque de misterio
- Audiencia: Cinéfilos, amantes del entretenimiento, usuarios de redes sociales
- Duración: Máximo 3 minutos por video

Reglas importantes:
1. NO reveles spoilers importantes de la trama
2. Mantén un tono dinámico y emocionante
3. Incluye elementos visuales y de audio específicos
4. Optimiza para la plataforma objetivo
5. Usa lenguaje natural y conversacional
6. Incluye call-to-actions para engagement

Responde SOLO con un JSON válido que contenga:
{
  "sections": [
    {
      "type": "intro|hook|plot|analysis|outro",
      "content": "texto del guion",
      "duration_seconds": 30,
      "visual_cues": ["indicación1", "indicación2"],
      "emotion": "excitement|suspense|drama|comedy|neutral",
      "emphasis_words": ["palabra1", "palabra2"]
    }
  ]
}
        `.trim();
    }
    
    /**
     * Parsea las secciones del guion desde el texto generado
     * @param {string} scriptText - Texto del guion
     * @param {Object} content - Información del contenido
     * @returns {Array} Secciones del guion
     */
    _parseScriptSections(scriptText, content) {
        try {
            // Intentar extraer JSON del texto
            const jsonMatch = scriptText.match(/\{[\s\S]*\}/);
            if (jsonMatch) {
                const data = JSON.parse(jsonMatch[0]);
                return data.sections || [];
            }
        } catch (error) {
            console.warn('Error parseando JSON del guion:', error);
        }
        
        // Fallback: parsear texto plano
        return this._parseTextSections(scriptText, content);
    }
    
    /**
     * Parsea secciones desde texto plano
     * @param {string} scriptText - Texto del guion
     * @param {Object} content - Información del contenido
     * @returns {Array} Secciones del guion
     */
    _parseTextSections(scriptText, content) {
        const sections = [];
        const lines = scriptText.split('\n').filter(line => line.trim());
        
        let currentSection = null;
        let currentContent = [];
        
        for (const line of lines) {
            const trimmedLine = line.trim();
            
            // Detectar inicio de nueva sección
            if (this._isSectionHeader(trimmedLine)) {
                // Guardar sección anterior
                if (currentSection && currentContent.length > 0) {
                    sections.push(this._createSectionFromText(
                        currentSection, 
                        currentContent.join('\n'),
                        content
                    ));
                }
                
                // Iniciar nueva sección
                currentSection = this._extractSectionType(trimmedLine);
                currentContent = [];
            } else if (currentSection && trimmedLine) {
                currentContent.push(trimmedLine);
            }
        }
        
        // Agregar última sección
        if (currentSection && currentContent.length > 0) {
            sections.push(this._createSectionFromText(
                currentSection,
                currentContent.join('\n'),
                content
            ));
        }
        
        // Si no se detectaron secciones, crear una sección general
        if (sections.length === 0) {
            sections.push(this._createSectionFromText(
                'plot',
                scriptText,
                content
            ));
        }
        
        return sections;
    }
    
    /**
     * Verifica si una línea es un encabezado de sección
     * @param {string} line - Línea a verificar
     * @returns {boolean} True si es encabezado
     */
    _isSectionHeader(line) {
        const sectionPatterns = [
            /^(?:INTRO|INTRODUCCIÓN)/i,
            /^(?:HOOK|GANCHO)/i,
            /^(?:PLOT|TRAMA|SINOPSIS)/i,
            /^(?:ANÁLISIS|ANALISIS)/i,
            /^(?:OUTRO|CIERRE|CONCLUSIÓN)/i
        ];
        
        return sectionPatterns.some(pattern => pattern.test(line));
    }
    
    /**
     * Extrae el tipo de sección de una línea
     * @param {string} line - Línea con encabezado
     * @returns {string} Tipo de sección
     */
    _extractSectionType(line) {
        if (/^(?:INTRO|INTRODUCCIÓN)/i.test(line)) return 'intro';
        if (/^(?:HOOK|GANCHO)/i.test(line)) return 'hook';
        if (/^(?:PLOT|TRAMA|SINOPSIS)/i.test(line)) return 'plot';
        if (/^(?:ANÁLISIS|ANALISIS)/i.test(line)) return 'analysis';
        if (/^(?:OUTRO|CIERRE|CONCLUSIÓN)/i.test(line)) return 'outro';
        return 'plot';
    }
    
    /**
     * Crea una sección desde texto plano
     * @param {string} sectionType - Tipo de sección
     * @param {string} content - Contenido de la sección
     * @param {Object} contentInfo - Información del contenido
     * @returns {Object} Sección del guion
     */
    _createSectionFromText(sectionType, content, contentInfo) {
        const wordCount = content.split(' ').length;
        const duration = Math.max(10, Math.round((wordCount / 150) * 60)); // 150 palabras por minuto
        
        const emotionMap = {
            intro: 'excitement',
            hook: 'suspense',
            plot: 'drama',
            analysis: 'neutral',
            outro: 'excitement'
        };
        
        return {
            type: sectionType,
            content: content,
            duration_seconds: duration,
            visual_cues: this._generateVisualCues(sectionType),
            emotion: emotionMap[sectionType] || 'neutral',
            emphasis_words: this._extractEmphasisWords(content)
        };
    }
    
    /**
     * Genera indicaciones visuales para una sección
     * @param {string} sectionType - Tipo de sección
     * @returns {Array} Indicaciones visuales
     */
    _generateVisualCues(sectionType) {
        const cues = {
            intro: ['Logo Cine Norte animado', 'Efecto de luces de reflector', 'Transición dinámica'],
            hook: ['Clip más impactante del tráiler', 'Texto en pantalla con título', 'Efecto de zoom dramático'],
            plot: ['Montaje de escenas clave', 'Texto descriptivo superpuesto', 'Transiciones suaves'],
            analysis: ['Split screen con comparaciones', 'Gráficos informativos', 'Efectos de partículas'],
            outro: ['Logo Cine Norte final', 'Call-to-action visual', 'Fade out elegante']
        };
        
        return cues[sectionType] || ['Elementos visuales dinámicos'];
    }
    
    /**
     * Extrae palabras de énfasis del texto
     * @param {string} text - Texto a analizar
     * @returns {Array} Palabras de énfasis
     */
    _extractEmphasisWords(text) {
        const emphasisPatterns = [
            /\b(?:increíble|espectacular|impresionante|sorprendente)\b/gi,
            /\b(?:nunca|siempre|definitivamente|absolutamente)\b/gi,
            /\b(?:¡.*!)\b/g,
            /\b(?:más|mejor|peor|único|especial)\b/gi
        ];
        
        const emphasisWords = [];
        for (const pattern of emphasisPatterns) {
            const matches = text.match(pattern);
            if (matches) {
                emphasisWords.push(...matches);
            }
        }
        
        return [...new Set(emphasisWords)].slice(0, 5);
    }
    
    /**
     * Genera sugerencias de títulos optimizados
     * @param {Object} content - Información del contenido
     * @param {string} targetPlatform - Plataforma objetivo
     * @returns {Array} Sugerencias de títulos
     */
    _generateTitleSuggestions(content, targetPlatform) {
        const baseTitle = content.title;
        const suggestions = [
            `¿Vale la pena ver ${baseTitle}? | Análisis Cine Norte`,
            `${baseTitle}: Todo lo que necesitas saber`,
            `Mi opinión sobre ${baseTitle} | Sin spoilers`,
            `${baseTitle} - Reseña completa en 3 minutos`,
            `¿${baseTitle} es tan buena como dicen? | Cine Norte`
        ];
        
        // Ajustar para TikTok (más cortos)
        if (targetPlatform === 'tiktok') {
            return [
                `${baseTitle} en 60 segundos`,
                `Mi veredicto: ${baseTitle}`,
                `${baseTitle} - ¿Sí o no?`,
                `Todo sobre ${baseTitle}`,
                `${baseTitle} - Sin spoilers`
            ];
        }
        
        return suggestions;
    }
    
    /**
     * Determina el estilo visual recomendado
     * @param {Object} content - Información del contenido
     * @param {string} style - Estilo del guion
     * @returns {string} Estilo visual
     */
    _determineVisualStyle(content, style) {
        if (content.genre.includes('Acción') || content.genre.includes('Ciencia Ficción')) {
            return 'high_energy';
        }
        if (content.genre.includes('Drama') || content.genre.includes('Romance')) {
            return 'cinematic';
        }
        if (content.genre.includes('Terror') || content.genre.includes('Suspenso')) {
            return 'dark_mysterious';
        }
        if (content.genre.includes('Comedia')) {
            return 'bright_playful';
        }
        return 'professional';
    }
    
    /**
     * Sugiere música de fondo
     * @param {Object} content - Información del contenido
     * @param {string} style - Estilo del guion
     * @returns {string} Sugerencia de música
     */
    _suggestMusic(content, style) {
        if (content.genre.includes('Acción')) return 'epic_action';
        if (content.genre.includes('Drama')) return 'emotional_drama';
        if (content.genre.includes('Terror')) return 'tense_horror';
        if (content.genre.includes('Comedia')) return 'light_comedy';
        return 'cinematic_ambient';
    }
    
    /**
     * Crea un guion de respaldo si falla la IA
     * @param {Object} content - Información del contenido
     * @param {string} targetPlatform - Plataforma objetivo
     * @returns {Object} Guion de respaldo
     */
    _createFallbackScript(content, targetPlatform) {
        const introText = `¡Hola cinéfilos! Bienvenidos a Cine Norte. Hoy analizamos ${content.title}.`;
        const plotText = `${content.overview.substring(0, 200)}...`;
        const outroText = '¿Qué opinas de esta película? Déjanos tu comentario y suscríbete para más análisis.';
        
        const sections = [
            {
                type: 'intro',
                content: introText,
                duration_seconds: 10,
                visual_cues: ['Logo Cine Norte'],
                emotion: 'excitement',
                emphasis_words: ['cinéfilos']
            },
            {
                type: 'plot',
                content: plotText,
                duration_seconds: 60,
                visual_cues: ['Montaje de escenas'],
                emotion: 'drama',
                emphasis_words: []
            },
            {
                type: 'outro',
                content: outroText,
                duration_seconds: 15,
                visual_cues: ['Call-to-action'],
                emotion: 'excitement',
                emphasis_words: ['comentario']
            }
        ];
        
        return {
            title: content.title,
            content: content,
            sections: sections,
            total_duration: 85,
            word_count: (introText + plotText + outroText).split(' ').length,
            target_platform: targetPlatform,
            hashtags: config.hashtags.slice(0, 10),
            title_suggestions: [`Análisis de ${content.title}`],
            visual_style: 'professional',
            music_suggestion: 'cinematic_ambient',
            raw_text: introText + '\n\n' + plotText + '\n\n' + outroText,
            generated_at: new Date().toISOString()
        };
    }
    
    /**
     * Exporta el guion a formato de texto
     * @param {Object} script - Guion generado
     * @returns {string} Contenido del archivo
     */
    exportScriptToText(script) {
        let content = `GUION CINE NORTE
================
Título: ${script.title}
Plataforma: ${script.target_platform}
Duración: ${script.total_duration} segundos
Palabras: ${script.word_count}

HASHTAGS:
${script.hashtags.join(', ')}

TÍTULOS SUGERIDOS:
${script.title_suggestions.map(title => `- ${title}`).join('\n')}

ESTILO VISUAL: ${script.visual_style}
MÚSICA: ${script.music_suggestion}

GUION:
======

`;
        
        script.sections.forEach((section, index) => {
            content += `
SECCIÓN ${index + 1}: ${section.type.toUpperCase()}
Duración: ${section.duration_seconds}s
Emoción: ${section.emotion}
Indicaciones visuales: ${section.visual_cues.join(', ')}

${section.content}

`;
        });
        
        return content;
    }
}

// Instancia global del generador
window.ScriptGenerator = ScriptGenerator;
window.scriptGenerator = new ScriptGenerator();
